### PR TITLE
Move packages out of deprecated pkg directory

### PR DESCRIPTION
**Pull request by beorn7**
Alternatively or in addition to #9477, I have taken action and just moved all packages out of the deprecated `pkg`.

This creates a new `model` directory and moves all data-model related
packages over there:
`exemplar labels relabel rulefmt textparse timestamp value`

All the others are more or less utilities and have been moved to `util`:
`gate logging modetimevfs pool runtime`

### CODE REVIEW

**Comment by beorn7** [comment]
Suggestion for the announcement mail:

"""
Dear Prometheus developers,

On 2021-08-22, the Prometheus dev summit has expressed a consensus to deprecate the `pkg` directory.

The `pkg` directory is used in some (but not all) repositories in the Prometheus GH org as a root directory for some (but not necessarily all) Go packages in a repository. The discussion of the pros and cons is nuanced (see the notes and recording of the dev summit), but the outcome is that we desire consistency between repositories and want to gravitate towards _not_ using the `pkg` directory. Please do not add more packages to any existing `pkg` directory, and move packages out of the directory at your convenience.

We plan to do the latter soon for prometheus/prometheus, and this mail intends to be a heads-up for those using the affected packages as libraries from outside the prometheus/prometheus repository.

The packages `gate`, `logging`, `modtimevfs`,  `pool`, and `runtime` will be moved to the (already existing) `util` directory.

The packages `exemplar`, `labels`, `relabel`, `rulefmt`, `textparse`, `timestamp`, and `value` will be moved to the (newly created) `model` directory.

It will happen in this pull request: https://github.com/prometheus/prometheus/pull/9478

Apologies for any hassle this may cause and thank you very much for your understanding,
"""

**Comment by roidelapluie** [comment]
:+1:

**Comment by roidelapluie** [comment]
> Mail has been set out: https://groups.google.com/g/prometheus-developers/c/M2wV7yc59Qs
> 
> How long should we wait before actually doing the move?

My suggestion is to do it when 2.31.0 is merged back to main.

**Comment by beorn7** [comment]
Fine with me.

**Comment by LeviHarrison** [comment]
I think the top-level package `template` might also fit better into `model`.

**Comment by beorn7** [comment]
> I think the top-level package template might also fit better into model.

Maybe. But let's do other moves after this is merged.

I will now recreate this PR with @roidelapluie 's script.

**Comment by beorn7** [comment]
No, wait, release-2.31 hasn't been merged back into main, so I'll wait for the release to happen.

**Comment by beorn7** [comment]
I have now recreated the move on top of main, using @roidelapluie 's script. But I had to do a few extra steps in the parser code and for MS Windows.

**Comment by beorn7** [comment]
Script hasn't done pkg/rulefmt/testdata/* move, either.

**Review by roidelapluie — APPROVED** [review]
LGTM

File: pkg/textparse/README.md
Comment (roidelapluie) [inline pkg/textparse/README.md:1]: Where is this file gone?

File: pkg/textparse/openmetricslex.l
Comment (roidelapluie) [inline pkg/textparse/openmetricslex.l:1]: Where is this file gone?

File: pkg/textparse/promtestdata.nometa.txt
Comment (roidelapluie) [inline pkg/textparse/promtestdata.nometa.txt:1]: Where is this file gone?

**Review by roidelapluie — CHANGES_REQUESTED** [review]
Some files seem to be missing (there is more than the one I highlighted)

**Comment by roidelapluie** [comment]
All missing files are from the same directory.

**Comment by beorn7** [comment]
Yeah, the script missed a bunch of files, and then I tried to find them all, but failed with a few. Thanks for your attention

**Comment by beorn7** [comment]
I hope I got them all now.

**Comment by roidelapluie** [comment]
168/168 files reviewed

I think so :)

**Review by roidelapluie — APPROVED** [review]
LGTM

**Comment by beorn7** [comment]
I did a whitespace commit to trigger the tests again. Some didn't run, for some reason.

**Comment by roidelapluie** [comment]
It was probably unluckily at the same time I added oauth restrictions, just before I authorize circle CI again.

**Review by roidelapluie — APPROVED** [review]

**Pull request merged**
